gen/zap.go

```diff
@@ -86,6 +86,27 @@ func (z *zapGenerator) zapEncoder(g Generator, spec compile.TypeSpec) string {
 	panic(root)
 }
 
+// zapTypedefHasGeneratedMarshaler defines if a typedef will have generated code
+// zapcore.ObjectMarshaler or zapcore.ArrayMarshaler.
+func (z *zapGenerator) zapTypedefHasGeneratedMarshaler(g Generator, spec compile.TypeSpec) bool {
+	return z.zapEncoder(g, spec) == "Array" || z.zapEncoder(g, spec) == "Object"
+}
+
+// zapTypedefGenerateMarshaler is similar to zapMarshaler but always converts a typedef
+// to its root value. This should be used when generating the code for the Zap
+// marshal implementation of the typedef.
+func (z *zapGenerator) zapTypedefGenerateMarshaler(g Generator, spec compile.TypeSpec, fieldValue string) (string, error) {
```

Comment:
I'm not sure why we need this separate function exposed in the template? Looks
like zapMarshaler can (and does) already do this if spec is a
compile.TypedefSpec.

Reply:
This `zapMarshaler` is used in the template that generates `MarshalLogObject`/`MarshalLogArray` for typedef, as well as in the template for other object's `MarshalLog...` function if they have typedef for a field. In one case you need the underlying one, in the other one you need to typeconvert to the underlying to avoid the typedef's generated `MarshalLog...` to call itself.

To avoid adding adding an extra parameter to `zapMarshaler` (something like `forceTypedefToRoot bool`), I made/exported a new function.